File: scripts/support/browser.py
"""Minimal Chromium DevTools transport and lifecycle support."""

from __future__ import annotations

import base64
import functools
import hashlib
import http.server
import json
import os
import secrets
import shutil
import signal
import socket
import struct
import subprocess
import tempfile
import threading
import time
import urllib.parse
import urllib.request
from pathlib import Path


class BrowserError(RuntimeError):
    pass


class QuietFileServer(http.server.SimpleHTTPRequestHandler):
    def log_message(self, *_: object) -> None:
        pass


class DevToolsSocket:
    def __init__(self, websocket_url: str) -> None:
        parsed = urllib.parse.urlsplit(websocket_url)
        self.socket = socket.create_connection((parsed.hostname, parsed.port), timeout=10)
        self.socket.settimeout(15)
        key = base64.b64encode(secrets.token_bytes(16)).decode("ascii")
        path = parsed.path + (f"?{parsed.query}" if parsed.query else "")
        request = (
            f"GET {path} HTTP/1.1\r\n"
            f"Host: {parsed.hostname}:{parsed.port}\r\n"
            "Upgrade: websocket\r\n"
            "Connection: Upgrade\r\n"
            f"Sec-WebSocket-Key: {key}\r\n"
            "Sec-WebSocket-Version: 13\r\n\r\n"
        ).encode("ascii")
        self.socket.sendall(request)
        response = self._read_headers()
        expected = base64.b64encode(
            hashlib.sha1((key + "258EAFA5-E914-47DA-95CA-C5AB0DC85B11").encode("ascii")).digest()
        ).decode("ascii")
        if "101" not in response.split("\r\n", 1)[0] or f"sec-websocket-accept: {expected}".lower() not in response.lower():
            raise BrowserError("Chromium DevTools WebSocket handshake failed.")
        self.next_id = 0

    def _read_headers(self) -> str:
        data = bytearray()
        while b"\r\n\r\n" not in data:
            chunk = self.socket.recv(1)
            if not chunk:
                raise BrowserError("Chromium closed the DevTools connection during handshake.")
            data.extend(chunk)
        return data.decode("iso-8859-1")

    def _read_exact(self, length: int) -> bytes:
        chunks = bytearray()
        while len(chunks) < length:
            chunk = self.socket.recv(length - len(chunks))
            if not chunk:
                raise BrowserError("Chromium closed the DevTools connection.")
            chunks.extend(chunk)
        return bytes(chunks)

    def _send_frame(self, opcode: int, payload: bytes) -> None:
        mask = secrets.token_bytes(4)
        length = len(payload)
        if length < 126:
            header = bytes((0x80 | opcode, 0x80 | length))
        elif length < 65536:
            header = bytes((0x80 | opcode, 0x80 | 126)) + struct.pack("!H", length)
        else:
            header = bytes((0x80 | opcode, 0x80 | 127)) + struct.pack("!Q", length)
        masked = bytes(value ^ mask[index % 4] for index, value in enumerate(payload))
        self.socket.sendall(header + mask + masked)

    def _receive_frame(self) -> tuple[int, bytes]:
        first, second = self._read_exact(2)
        opcode = first & 0x0F
        masked = bool(second & 0x80)
        length = second & 0x7F
        if length == 126:
            length = struct.unpack("!H", self._read_exact(2))[0]
        elif length == 127:
            length = struct.unpack("!Q", self._read_exact(8))[0]
        mask = self._read_exact(4) if masked else b""
        payload = self._read_exact(length)
        if masked:
            payload = bytes(value ^ mask[index % 4] for index, value in enumerate(payload))
        return opcode, payload

    def command(self, method: str, params: dict | None = None) -> dict:
        self.next_id += 1
        message_id = self.next_id
        payload = {"id": message_id, "method": method}
        if params:
            payload["params"] = params
        self._send_frame(1, json.dumps(payload, separators=(",", ":")).encode("utf-8"))
        fragments: list[bytes] = []
        while True:
            opcode, message = self._receive_frame()
            if opcode == 8:
                raise BrowserError("Chromium closed the DevTools connection before responding.")
            if opcode == 9:
                self._send_frame(10, message)
                continue
            if opcode == 1:
                fragments = [message]
            elif opcode == 0:
                fragments.append(message)
            else:
                continue
            try:
                response = json.loads(b"".join(fragments))
            except json.JSONDecodeError:
                continue
            if response.get("id") != message_id:
                continue
            if "error" in response:
                raise BrowserError(f"Chromium command {method} failed: {response['error'].get('message', response['error'])}")
            return response.get("result", {})

    def close(self) -> None:
        try:
            self._send_frame(8, b"")
        finally:
            self.socket.close()


def http_json(url: str, method: str = "GET") -> dict:
    request = urllib.request.Request(url, method=method)
    with urllib.request.urlopen(request, timeout=2) as response:
        return json.load(response)


def evaluate(browser: DevToolsSocket, expression: str) -> object:
    result = browser.command(
        "Runtime.evaluate",
        {"expression": expression, "awaitPromise": True, "returnByValue": True},
    )
    if "exceptionDetails" in result:
        raise BrowserError(f"Browser evaluation failed: {result['exceptionDetails'].get('text', 'unknown exception')}")
    return result.get("result", {}).get("value")


def wait_for_complete(browser: DevToolsSocket, route: str, label: str | None = None) -> None:
    deadline = time.monotonic() + 15
    while time.monotonic() < deadline:
        if evaluate(browser, "[document.readyState, location.pathname]") == ["complete", route]:
            time.sleep(0.15)
            return
        time.sleep(0.05)
    raise BrowserError(f"Timed out rendering {label or route} in Chromium.")


def set_viewport(browser: DevToolsSocket, width: int, height: int) -> None:
    browser.command(
        "Emulation.setDeviceMetricsOverride",
        {"width": width, "height": height, "deviceScaleFactor": 1, "mobile": False},
    )


def open_page(debug_url: str, origin: str, route: str) -> DevToolsSocket:
    target_url = origin + route
    endpoint = debug_url + "/json/new?" + urllib.parse.quote(target_url, safe=":/?=&")
    try:
        target = http_json(endpoint, method="PUT")
    except Exception as error:
        raise BrowserError(f"Could not open {route} in Chromium: {error}") from error
    return DevToolsSocket(target["webSocketDebuggerUrl"])


class ChromiumSession:
    def __init__(self, document_root: Path) -> None:
        self.document_root = document_root
        self.server: http.server.ThreadingHTTPServer | None = None
        self.server_thread: threading.Thread | None = None
        self.profile: tempfile.TemporaryDirectory[str] | None = None
        self.process: subprocess.Popen[bytes] | None = None
        self.origin = ""
        self.debug_url = ""

    def __enter__(self) -> ChromiumSession:
        if not self.document_root.is_dir():
            raise BrowserError(f"Missing site directory: {self.document_root}")
        chromium = shutil.which("chromium") or shutil.which("chromium-browser") or shutil.which("google-chrome")
        if not chromium:
            raise BrowserError("Missing Chromium; install chromium to run browser acceptance checks.")

        self.server = http.server.ThreadingHTTPServer(
            ("127.0.0.1", 0),
            functools.partial(QuietFileServer, directory=str(self.document_root)),
        )
        self.server_thread = threading.Thread(target=self.server.serve_forever, daemon=True)
        self.server_thread.start()
        self.origin = f"http://127.0.0.1:{self.server.server_port}"

        with socket.socket() as port_socket:
            port_socket.bind(("127.0.0.1", 0))
            debug_port = port_socket.getsockname()[1]
        self.profile = tempfile.TemporaryDirectory(prefix="chromium-")
        self.process = subprocess.Popen(
            [
                chromium,
                "--headless=new",
                f"--remote-debugging-port={debug_port}",
                "--remote-debugging-address=127.0.0.1",
                f"--user-data-dir={self.profile.name}",
                "--no-first-run",
                "--no-default-browser-check",
                "--disable-background-networking",
                "--disable-component-update",
                "--disable-sync",
                "--metrics-recording-only",
                "about:blank",
            ],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
            start_new_session=True,
        )
        self.debug_url = f"http://127.0.0.1:{debug_port}"
        deadline = time.monotonic() + 15
        while True:
            try:
                http_json(self.debug_url + "/json/version")
                return self
            except Exception:
                if time.monotonic() >= deadline:
                    self.close()
                    raise BrowserError("Chromium did not expose its DevTools endpoint within 15 seconds.")
                time.sleep(0.05)

    def open_page(self, route: str) -> DevToolsSocket:
        return open_page(self.debug_url, self.origin, route)

    def close(self) -> None:
        if self.process:
            try:
                os.killpg(self.process.pid, signal.SIGTERM)
            except ProcessLookupError:
                pass
            if self.process.poll() is None:
                try:
                    self.process.wait(timeout=5)
                except subprocess.TimeoutExpired:
                    os.killpg(self.process.pid, signal.SIGKILL)
                    self.process.wait(timeout=5)
        if self.profile:
            self.profile.cleanup()
        if self.server:
            self.server.shutdown()
            self.server.server_close()

    def __exit__(self, *_: object) -> None:
        self.close()

File: scripts/support/browser.mjs
import { createServer } from "node:http";
import { access, mkdtemp, readFile, rm, stat } from "node:fs/promises";
import { constants } from "node:fs";
import { spawn } from "node:child_process";
import { tmpdir } from "node:os";
import { delimiter, extname, resolve, sep } from "node:path";
import { setTimeout as sleep } from "node:timers/promises";

const MIME_TYPES = new Map([
  [".css", "text/css; charset=utf-8"], [".html", "text/html; charset=utf-8"],
  [".js", "text/javascript; charset=utf-8"], [".json", "application/json; charset=utf-8"],
  [".svg", "image/svg+xml"], [".wasm", "application/wasm"], [".xml", "application/xml"],
  [".png", "image/png"], [".webp", "image/webp"], [".woff", "font/woff"], [".woff2", "font/woff2"],
]);

export class BrowserError extends Error {}

function fail(message) { throw new BrowserError(message); }

async function serve(root, request, response) {
  try {
    const pathname = decodeURIComponent(new URL(request.url, "http://localhost").pathname);
    let file = resolve(root, "." + pathname);
    if (file !== root && !file.startsWith(root + sep)) return response.writeHead(403).end();
    if ((await stat(file)).isDirectory()) file = resolve(file, "index.html");
    const body = await readFile(file);
    response.writeHead(200, { "Content-Type": MIME_TYPES.get(extname(file)) ?? "application/octet-stream" });
    response.end(request.method === "HEAD" ? undefined : body);
  } catch {
    response.writeHead(404).end();
  }
}

function listen(server) {
  return new Promise((resolve, reject) => {
    server.once("error", reject);
    server.listen(0, "127.0.0.1", () => {
      server.off("error", reject);
      resolve(server.address().port);
    });
  });
}

async function findCommand(names) {
  for (const directory of (process.env.PATH ?? "").split(delimiter)) {
    for (const name of names) {
      const candidate = resolve(directory, name);
      try { await access(candidate, constants.X_OK); return candidate; } catch {}
    }
  }
  return null;
}

async function json(url, options) {
  const response = await fetch(url, { ...options, signal: AbortSignal.timeout(2_000) });
  if (!response.ok) fail(`Chromium DevTools request failed: ${response.status} ${response.statusText}.`);
  return response.json();
}

class DevToolsSocket {
  #socket;
  #nextId = 0;
  #pending = new Map();
  #events = [];
  #waiters = [];

  static async connect(url) {
    const socket = new WebSocket(url);
    await new Promise((resolve, reject) => {
      const timer = setTimeout(() => reject(new BrowserError("Chromium DevTools WebSocket handshake timed out.")), 10_000);
      socket.addEventListener("open", () => { clearTimeout(timer); resolve(); }, { once: true });
      socket.addEventListener("error", () => { clearTimeout(timer); reject(new BrowserError("Chromium DevTools WebSocket handshake failed.")); }, { once: true });
    });
    return new DevToolsSocket(socket);
  }

  constructor(socket) {
    this.#socket = socket;
    socket.addEventListener("message", event => this.#receive(JSON.parse(event.data)));
    socket.addEventListener("close", () => this.#closePending());
    socket.addEventListener("error", () => this.#closePending());
  }

  #closePending() {
    for (const { reject, timer } of this.#pending.values()) {
      clearTimeout(timer);
      reject(new BrowserError("Chromium closed the DevTools connection before responding."));
    }
    this.#pending.clear();
    for (const { reject, timer } of this.#waiters.splice(0)) {
      clearTimeout(timer);
      reject(new BrowserError("Chromium closed the DevTools connection before emitting an event."));
    }
  }

  #receive(message) {
    if ("id" in message) {
      const pending = this.#pending.get(message.id);
      if (!pending) return;
      this.#pending.delete(message.id);
      clearTimeout(pending.timer);
      if (message.error) pending.reject(new BrowserError(`Chromium command ${pending.method} failed: ${message.error.message ?? JSON.stringify(message.error)}`));
      else pending.resolve(message.result ?? {});
      return;
    }
    if (!message.method) return;
    const index = this.#waiters.findIndex(waiter => waiter.method === message.method);
    if (index >= 0) {
      const [{ resolve, timer }] = this.#waiters.splice(index, 1);
      clearTimeout(timer);
      resolve(message.params ?? {});
    } else this.#events.push(message);
  }

  command(method, params) {
    const id = ++this.#nextId;
    return new Promise((resolve, reject) => {
      const timer = setTimeout(() => {
        this.#pending.delete(id);
        reject(new BrowserError(`Timed out waiting for Chromium command ${method}.`));
      }, 15_000);
      this.#pending.set(id, { method, resolve, reject, timer });
      this.#socket.send(JSON.stringify(params ? { id, method, params } : { id, method }));
    });
  }

  waitForEvent(method, timeout = 15_000) {
    const index = this.#events.findIndex(event => event.method === method);
    if (index >= 0) return Promise.resolve(this.#events.splice(index, 1)[0].params ?? {});
    return new Promise((resolve, reject) => {
      const timer = setTimeout(() => {
        const index = this.#waiters.findIndex(waiter => waiter.resolve === resolve);
        if (index >= 0) this.#waiters.splice(index, 1);
        reject(new BrowserError(`Timed out waiting for Chromium event ${method}.`));
      }, timeout);
      this.#waiters.push({ method, resolve, reject, timer });
    });
  }

  close() { this.#socket.close(); }
}

export async function evaluate(browser, expression) {
  const result = await browser.command("Runtime.evaluate", { expression, awaitPromise: true, returnByValue: true });
  if (result.exceptionDetails) fail(`Browser evaluation failed: ${result.exceptionDetails.text ?? "unknown exception"}`);
  return result.result?.value;
}

export async function waitForComplete(browser, route, label = route) {
  const deadline = Date.now() + 15_000;
  while (Date.now() < deadline) {
    if (JSON.stringify(await evaluate(browser, "[document.readyState, location.pathname]")) === JSON.stringify(["complete", route])) {
      await sleep(150);
      return;
    }
    await sleep(50);
  }
  fail(`Timed out rendering ${label} in Chromium.`);
}

export function setViewport(browser, width, height) {
  return browser.command("Emulation.setDeviceMetricsOverride", { width, height, deviceScaleFactor: 1, mobile: false });
}

export class ChromiumSession {
  constructor(documentRoot) { this.documentRoot = documentRoot; }

  async open() {
    try { await stat(this.documentRoot); } catch { fail(`Missing site directory: ${this.documentRoot}`); }
    const chromium = await findCommand(["chromium", "chromium-browser", "google-chrome"]);
    if (!chromium) fail("Missing Chromium; install chromium to run browser acceptance checks.");
    this.server = createServer((request, response) => { void serve(this.documentRoot, request, response); });
    const port = await listen(this.server);
    this.origin = `http://127.0.0.1:${port}`;
    this.profile = await mkdtemp(resolve(tmpdir(), "chromium-"));
    const debugServer = createServer();
    const debugPort = await listen(debugServer);
    await new Promise(resolve => debugServer.close(resolve));
    this.process = spawn(chromium, [
      "--headless=new", `--remote-debugging-port=${debugPort}`, "--remote-debugging-address=127.0.0.1",
      `--user-data-dir=${this.profile}`, "--no-first-run", "--no-default-browser-check",
      "--disable-background-networking", "--disable-component-update", "--disable-sync", "--metrics-recording-only", "about:blank",
    ], { detached: true, stdio: "ignore" });
    this.debugUrl = `http://127.0.0.1:${debugPort}`;
    const deadline = Date.now() + 15_000;
    while (Date.now() < deadline) {
      try { await json(this.debugUrl + "/json/version"); return this; } catch { await sleep(50); }
    }
    await this.close();
    fail("Chromium did not expose its DevTools endpoint within 15 seconds.");
  }

  async openPage(route) {
    try {
      const target = await json(this.debugUrl + "/json/new?" + encodeURIComponent(this.origin + route), { method: "PUT" });
      return DevToolsSocket.connect(target.webSocketDebuggerUrl);
    } catch (error) {
      throw new BrowserError(`Could not open ${route} in Chromium: ${error.message ?? error}`);
    }
  }

  async close() {
    if (this.process && this.process.exitCode === null) {
      try { process.kill(-this.process.pid, "SIGTERM"); } catch {}
      await Promise.race([new Promise(resolve => this.process.once("exit", resolve)), sleep(5_000)]);
      if (this.process.exitCode === null) { try { process.kill(-this.process.pid, "SIGKILL"); } catch {} }
    }
    if (this.profile) await rm(this.profile, { recursive: true, force: true });
    if (this.server) await new Promise(resolve => this.server.close(resolve));
  }
}
